ddtrace/context.py

```diff
@@ -1,219 +1,31 @@
-import logging
-import threading
-
-from .constants import HOSTNAME_KEY, SAMPLING_PRIORITY_KEY, ORIGIN_KEY, LOG_SPAN_KEY
 from .internal.logger import get_logger
-from .internal import hostname
-from .settings import config
-from .utils.formats import asbool, get_env
 
 log = get_logger(__name__)
 
 
 class Context(object):
-    """
-    Context is used to keep track of a hierarchy of spans for the current
-    execution flow. During each logical execution, the same ``Context`` is
-    used to represent a single logical trace, even if the trace is built
-    asynchronously.
-
-    A single code execution may use multiple ``Context`` if part of the execution
-    must not be related to the current tracing. As example, a delayed job may
-    compose a standalone trace instead of being related to the same trace that
-    generates the job itself. On the other hand, if it's part of the same
-    ``Context``, it will be related to the original trace.
-
-    This data structure is thread-safe.
+    """Represents a remote trace and active span in order to perform distributed
+    tracing.
     """
 
-    _partial_flush_enabled = asbool(get_env("tracer", "partial_flush_enabled", default=False))
-    _partial_flush_min_spans = int(get_env("tracer", "partial_flush_min_spans", default=500))
+    __slots__ = [
+        "trace_id",
+        "span_id",
+        "sampling_priority",
+        "_dd_origin",
+    ]
 
     def __init__(self, trace_id=None, span_id=None, sampling_priority=None, _dd_origin=None):
-        """
-        Initialize a new thread-safe ``Context``.
-
-        :param int trace_id: trace_id of parent span
-        :param int span_id: span_id of parent span
-        """
-        self._trace = []
-        self._finished_spans = 0
-        self._current_span = None
-        self._lock = threading.Lock()
-
-        self._parent_trace_id = trace_id
-        self._parent_span_id = span_id
-        self._sampling_priority = sampling_priority
+        # type (Optional[int], Optional[int], Optional[int], Optional[str]) -> None
```

Comment:
IMO it shouldn't even be public as it's an implementation detail. A user doesn't need to know or create context objects as it should be handled by propagators.

Reply:
Well, right now there is a chance people are interacting with contexts right? so how close to the original class do we need to keep this?

like I said, maybe NamedTuple would be better here, smaller footprint, more performant, etc ?

Reply:
The only place where they might is in distributed tracing in which case the usage should be just these same fields. Yes we can experiment with NamedTuple to see if we get any gains

Reply:
We are using `Context`. We have a custom propagator that uses different headers than `x-datadog-trace-id`. Do you have another recommendation for this use-case?

FYI: the last PR to `Context` was also a breaking change for a re-name of the attribute. 
https://github.com/DataDog/dd-trace-py/pull/1982

Reply:
@r-kells do you mind sharing a bit more about your usage of `Context`? Having your use-case would be really helpful in making sure we're not breaking anything 🙂 

The propagation API should not change at all (this change is mainly pulling _everything else_ out of `Context`).

Reply:
@Kyle-Verhoog Thanks for taking the time to double check our use-case.

We essentially have a [custom propagator](https://ddtrace.readthedocs.io/en/stable/advanced_usage.html#custom)

I double checked all our usages. We just have to able to safely **set** and **get** the attributes on this object. We never used any of the methods that are removed 👍 , just passed the newly created context to `activate`. 

It is being used inside a gRPC interceptor (within a threadpool).

I don't exactly know how the global tracer object stores references to different contexts (multiple traces will be in flight on any given server). So I'm not 100% if the now delete locks are needed or not.

Reply:
Thanks for checking @r-kells! We'd actually love to chat more about your usage of the library please feel free reach out to me on https://chat.datadoghq.com/ 😄 

The thread safety shouldn't be an issue as the context objects will now be thread/task-local whenever used in concurrent environments. (The lock was previously required for the trace building and because context objects could be shared across threads/tasks)